File: 110_balanced_binary_tree/isBalanced.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root:
            left = self.getHeight(root.left, 0)
            right = self.getHeight(root.right, 0)

            skew = abs(left - right)

            return skew <= 1 and self.isBalanced(root.left) and self.isBalanced(root.right) 

        return True

        
    def getHeight(self, root, height):
        if not root:
            return 0
        lHeight = self.getHeight(root.left, height)
        rHeight = self.getHeight(root.right, height)

        return max(lHeight, rHeight) + 1
        

File: 110_balanced_binary_tree/isBalanced.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
	if root != nil {
		left := getHeight(root.Left)
		right := getHeight(root.Right)
		skew := abs(left, right)

		if !isBalanced(root.Left) {
			return false
		}
		if !isBalanced(root.Right) {
			return false
		}
		if skew > 1 {
			return false
		}

		return true
	}

	return true
}

func getHeight(root *TreeNode) int {
	if root != nil {
		left := getHeight(root.Left)
		right := getHeight(root.Right)

		return max(left, right) + 1
	}

	return 0
}

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}

func abs(a, b int) int {
	c := a - b
	if c < 0 {
		return -c
	}

	return c
}
